File: watchface/src/c/display.c
#include "display.h"
#include <string.h>
#include "zone_column.h"
// Styles: 0 Span, 2 broad, 4 Chamfer, 5-8 system fonts, 9 Leco Delta. Retired:
// 1 triangular segments (migrates to Chamfer), 3 LCD (migrates to broad).
// Accept saved experimental packets so their display choice can be migrated.
// Version 2 packets carry the map background in byte 3; version 1 used it for
// the retired framing experiment.
static bool options_valid(const uint8_t *p){
  return !(p[2]&~0xfe)&&((p[2]>>2)&3)<ZONE_TIMES_COUNT&&((p[2]>>4)&3)<ZONE_POSITION_COUNT&&p[3]<MAP_BACKGROUND_COUNT;
}
// Version 3 byte 3 also carries the map time size (bits 2-4).
static bool v3_options_valid(const uint8_t *p){
  return !(p[2]&~0xfe)&&((p[2]>>2)&3)<ZONE_TIMES_COUNT&&((p[2]>>4)&3)<ZONE_POSITION_COUNT&&
    (p[3]&3)<MAP_BACKGROUND_COUNT&&((p[3]>>2)&7)<MAP_TIME_SIZE_CODES&&!(p[3]&~0x7f);
}
bool display_valid(const uint8_t *p,size_t length){
  if(!p||p[1]>9)return false;
  // Version 3 adds power and motion (power.h) in bytes 4-7.
  if(length==DISPLAY_SIZE)return p[0]==3&&v3_options_valid(p)&&!(p[4]&~0x7f)&&p[5]<24&&p[6]<24&&(p[7]==5||p[7]==10||p[7]==20||p[7]==30);
  if(length!=DISPLAY_LEGACY_SIZE)return false;
  // Version 2 byte 2: bit 1 no leading zero, bits 2-3 when place times also
  // show outside the panel, bits 4-5 where, bit 6 map times may turn, bit 7 the nameplate.
  if(p[0]==2)return options_valid(p);
  return p[0]==1&&p[2]<=3&&p[3]<64&&(p[3]&3)<3&&((p[3]&3)||!p[3]);
}
bool display_normalize(uint8_t out[DISPLAY_SIZE],const uint8_t *data,size_t length){
  if(!out||!display_valid(data,length))return false;
  uint8_t in[DISPLAY_SIZE]={0,0,0,0,0,22,7,10}; // older packets: default power and motion
  memcpy(in,data,length);memcpy(out,in,DISPLAY_SIZE);
  if(out[1]==3)out[1]=2;
  if(out[1]==1)out[1]=4;
  // Version 1: bit 0 (unlit triangles) is retired and byte 3 was framing.
  if(out[0]==1){out[2]&=2;out[3]=0;}
  out[0]=3;
  return true;
}

File: shared/display.js
// Numeral styles and the DISPLAY packet. Codes 5-8 are Pebble system fonts,
// 9 is Leco Delta (shared/system-clock.js).
export const DISPLAY_STYLES=['span','broad','chamfer','leco','bitham-bold','bitham-light','bitham-medium','leco-delta'];
// Wire codes. 1 was the retired triangular seven-segment style, which the watch
// migrates to Chamfer; 3 was the retired LCD style, migrated to broad.
export const DISPLAY_CODES=Object.freeze({span:0,broad:2,chamfer:4,leco:5,'bitham-bold':6,'bitham-light':7,'bitham-medium':8,'leco-delta':9});
import {MAP_BACKGROUNDS} from './map-background.js';
import {ZONE_TIMES,ZONE_POSITIONS} from './zone-column.js';
import {encodePower} from './power.js';
// Version 3: [3, style, options, map background, power bits, night start,
// night end, low battery]. Byte 2: bit 1 no leading zero (clear by default), bits 2-3
// index ZONE_TIMES, bits 4-5 ZONE_POSITIONS, bit 6 lets map times turn 90°, bit
// 7 shows the Dymaxion nameplate; bit 0 is retired. Byte 3 indexes
// MAP_BACKGROUNDS. Bytes 4-6 are power and motion (shared/power.js). The watch
// still loads 4-byte version 1 and 2 packets, with default power and motion.
export function encodeDisplay(settings){
  const style=DISPLAY_CODES[settings.clockDisplay],background=MAP_BACKGROUNDS.indexOf(settings.mapBackground??'none'),zones=ZONE_TIMES.indexOf(settings.zoneTimes??'panel');
  if(style===undefined)throw new Error('Invalid clock display.');
  if(background<0)throw new Error('Invalid map background.');
  if(zones<0)throw new Error('Invalid place-time placement.');
  const position=ZONE_POSITIONS.indexOf(settings.zonePosition??'left');
  if(position<0)throw new Error('Invalid place-time position.');
  return new Uint8Array([3,style,(settings.leadingZero===false?2:0)|zones<<2|position<<4|(settings.mapTimesTurn?64:0)|(settings.nameplate?128:0),background,...encodePower(settings.power)]);
}
